File: src/api/mods.ts
import type {
  Category,
  ModProject,
  ModUpdate,
  ModVersion,
  ProjectDetails,
  ProjectKind,
  ProviderId,
  ResolvedInstallPlan,
  SearchQuery,
  SearchResult,
} from '@/types/mod';
import { MOCK_PROJECTS } from '@/mocks/data';
import { ipc, ipcUnit, mocked, shouldMock } from './shared';
import { t } from '@/lib/i18n';

/** Providers the backend considers usable (CurseForge needs an API key). */
export function availableProviders(): Promise<ProviderId[]> {
  if (shouldMock()) return mocked<ProviderId[]>(['modrinth']);
  return ipc<ProviderId[]>('available_providers');
}

export function searchProjects(query: SearchQuery): Promise<SearchResult> {
  if (shouldMock()) {
    const text = query.text.trim().toLowerCase();
    const hits =
      text === ''
        ? MOCK_PROJECTS
        : MOCK_PROJECTS.filter(
            (project) =>
              project.name.toLowerCase().includes(text) ||
              project.summary.toLowerCase().includes(text),
          );
    return mocked<SearchResult>({ hits, totalHits: hits.length, offset: 0 }, 260);
  }
  return ipc<SearchResult>('search_projects', { query });
}

export function listCategories(provider: ProviderId, kind: ProjectKind): Promise<Category[]> {
  if (shouldMock()) {
    return mocked<Category[]>([
      { id: 'optimization', name: 'Optimization' },
      { id: 'utility', name: 'Utility' },
      { id: 'decoration', name: 'Decoration' },
    ]);
  }
  return ipc<Category[]>('list_categories', { provider, kind });
}

/** The full description page of a project. */
export function projectDetails(provider: ProviderId, projectId: string): Promise<ProjectDetails> {
  if (shouldMock()) {
    const project = MOCK_PROJECTS.find((item) => item.projectId === projectId) ?? MOCK_PROJECTS[0];
    if (project === undefined) return Promise.reject(new Error('no fixtures'));
    return mocked<ProjectDetails>(
      {
        project,
        body: [
          `## ${project.name}`,
          '',
          t`${project.summary} Описание из фикстур: **жирный**, *курсив*, \`код\` и [ссылка](https://modrinth.com).`,
          '',
          t`- Первый пункт`,
          t`- Второй пункт`,
          '',
          t`> Цитата автора.`,
          '',
          '<center><iframe src="https://www.youtube-nocookie.com/embed/dQw4w9WgXcQ"></iframe></center>',
          '',
          '<script>alert(1)</script><img src="x" onerror="alert(2)">',
        ].join('\n'),
        bodyFormat: 'markdown',
        gallery: [],
        links: [
          { kind: 'page', label: '', url: project.pageUrl ?? 'https://modrinth.com' },
          { kind: 'source', label: '', url: 'https://github.com' },
          { kind: 'donation', label: 'Ko-fi', url: 'https://ko-fi.com' },
        ],
        gameVersions: ['1.20.1', '1.20.4', '1.21.1'],
        loaders: ['fabric', 'quilt'],
        publishedAt: '2023-01-01T00:00:00Z',
      },
      300,
    );
  }
  return ipc<ProjectDetails>('project_details', { provider, projectId });
}

/** Content kinds that install into an instance (modpacks create one instead). */
export type ContentKind = Exclude<ProjectKind, 'modpack'>;

export function installedProjects(
  instanceId: string,
  provider: ProviderId,
  kind: ContentKind,
): Promise<string[]> {
  if (shouldMock()) return mocked<string[]>(kind === 'mod' ? ['AANobbMI'] : []);
  return ipc<string[]>('installed_projects', { instanceId, provider, kind });
}

function mockVersion(project: ModProject): ModVersion {
  return {
    provider: project.provider,
    projectId: project.projectId,
    versionId: `${project.projectId}-v`,
    name: project.name,
    versionNumber: '1.0.0',
    fileName: `${project.slug}-1.0.0.jar`,
    sizeBytes: 1_200_000,
    sha1: null,
    downloadUrl: 'https://example.invalid',
    gameVersions: ['1.20.4'],
    loaders: ['fabric'],
    releaseType: 'release',
    publishedAt: new Date().toISOString(),
    dependencies: [],
  };
}

/**
 * Versions for the picker, newest first. Only ones that fit the instance
 * unless `anyGameVersion` lifts the Minecraft version filter.
 */
export function listVersions(
  instanceId: string,
  provider: ProviderId,
  projectId: string,
  kind: ContentKind,
  anyGameVersion: boolean,
): Promise<ModVersion[]> {
  if (shouldMock()) {
    const project = MOCK_PROJECTS.find((item) => item.projectId === projectId) ?? MOCK_PROJECTS[0];
    if (project === undefined) return mocked<ModVersion[]>([]);
    const base = mockVersion(project);
    return mocked<ModVersion[]>(
      [
        { ...base, versionId: 'v3', versionNumber: '1.2.0-beta.1', releaseType: 'beta' },
        { ...base, versionId: `${projectId}-v`, versionNumber: '1.1.0' },
        {
          ...base,
          versionId: 'v1',
          versionNumber: '1.0.0',
          gameVersions: anyGameVersion ? ['1.19.4'] : base.gameVersions,
          publishedAt: '2024-03-01T00:00:00Z',
        },
      ],
      300,
    );
  }
  return ipc<ModVersion[]>('list_versions', {
    instanceId,
    provider,
    projectId,
    kind,
    anyGameVersion,
  });
}

/** `versionId` installs that exact version instead of the best fit. */
export function resolveInstall(
  instanceId: string,
  provider: ProviderId,
  projectId: string,
  kind: ContentKind,
  versionId: string | null = null,
): Promise<ResolvedInstallPlan> {
  if (shouldMock()) {
    const project = MOCK_PROJECTS.find((item) => item.projectId === projectId) ?? MOCK_PROJECTS[0];
    const api = MOCK_PROJECTS.find((item) => item.slug === 'fabric-api');
    if (project === undefined) return Promise.reject(new Error('no fixtures'));
    const dependencies = api !== undefined && api !== project ? [mockVersion(api)] : [];
    return mocked<ResolvedInstallPlan>(
      {
        primary: mockVersion(project),
        dependencies,
        unresolved: [],
        totalBytes: 1_200_000 * (1 + dependencies.length),
      },
      500,
    );
  }
  return ipc<ResolvedInstallPlan>('resolve_install', {
    instanceId,
    provider,
    projectId,
    kind,
    versionId,
  });
}

export function installPlan(
  instanceId: string,
  kind: ContentKind,
  plan: ResolvedInstallPlan,
): Promise<void> {
  if (shouldMock()) return mocked(undefined, 900);
  return ipcUnit('install_plan', { instanceId, kind, plan });
}

/** Deletes a resource pack or shader pack (mods go through `removeMod`). */
export function removeContent(
  instanceId: string,
  kind: ContentKind,
  fileName: string,
): Promise<void> {
  if (shouldMock()) return mocked(undefined, 100);
  return ipcUnit('remove_content', { instanceId, kind, fileName });
}

export function checkUpdates(instanceId: string, kind: ContentKind): Promise<ModUpdate[]> {
  if (shouldMock()) {
    if (kind !== 'mod') return mocked<ModUpdate[]>([], 500);
    const sodium = MOCK_PROJECTS.find((item) => item.slug === 'sodium');
    if (sodium === undefined) return mocked<ModUpdate[]>([]);
    return mocked<ModUpdate[]>(
      [
        {
          fileName: 'sodium-fabric-0.6.5.jar',
          currentVersion: '0.6.5',
          latest: { ...mockVersion(sodium), versionNumber: '0.6.7' },
        },
      ],
      700,
    );
  }
  return ipc<ModUpdate[]>('check_updates', { instanceId, kind });
}

export function applyUpdates(
  instanceId: string,
  kind: ContentKind,
  updates: readonly ModUpdate[],
): Promise<void> {
  if (shouldMock()) return mocked(undefined, 900);
  return ipcUnit('apply_updates', { instanceId, kind, updates });
}

export interface OptimizeMod {
  readonly projectId: string;
  readonly name: string;
  readonly about: string;
  /** Null when there is no version for this Minecraft version and loader. */
  readonly versionNumber: string | null;
  readonly installed: boolean;
}

export interface OptimizePlan {
  readonly mods: readonly OptimizeMod[];
  readonly currentMemoryMb: number;
  readonly recommendedMemoryMb: number;
  readonly currentJvmArgs: string;
  readonly recommendedJvmArgs: string;
}

export function optimizePlan(instanceId: string): Promise<OptimizePlan> {
  if (shouldMock()) {
    return mocked<OptimizePlan>(
      {
        mods: [
          { projectId: 'AANobbMI', name: 'Sodium', about: t`Новый движок рендера — FPS выше в разы`, versionNumber: '0.6.5', installed: true },
          { projectId: 'gvQqBUqZ', name: 'Lithium', about: t`Быстрее игровая логика: мобы, редстоун, чанки`, versionNumber: '0.14.3', installed: false },
          { projectId: 'uXXizFIs', name: 'FerriteCore', about: t`Заметно меньше расход оперативной памяти`, versionNumber: '7.0.3', installed: false },
          { projectId: 'nmDcB62a', name: 'ModernFix', about: t`Быстрее загрузка игры и миров`, versionNumber: null, installed: false },
        ],
        currentMemoryMb: 4096,
        recommendedMemoryMb: 6144,
        currentJvmArgs: '-XX:+UseG1GC',
        recommendedJvmArgs: '-XX:+UseG1GC -XX:MaxGCPauseMillis=200',
      },
      500,
    );
  }
  return ipc<OptimizePlan>('optimize_plan', { instanceId });
}

export function applyOptimize(
  instanceId: string,
  projectIds: readonly string[],
  memoryMb: number | null,
  jvmArgs: string | null,
): Promise<void> {
  if (shouldMock()) return mocked(undefined, 900);
  return ipcUnit('apply_optimize', { instanceId, projectIds, memoryMb, jvmArgs });
}

File: src-tauri/src/commands/mods.rs
use std::sync::Arc;

use tauri::State;

use crate::error::{ErrorKind, LauncherError, Result};
use crate::instances::{self, ModLoader};
use crate::mods::index::ModIndex;
use crate::mods::install::{self, ModUpdate};
use crate::mods::{
    self, resolve, Category, ModVersion, ProjectDetails, ProjectKind, ProviderId,
    ResolvedInstallPlan, SearchQuery, SearchResult, Target,
};
use crate::state::AppState;
use crate::tasks::{Progress, TaskKind};

/// What content of `kind` must support to go into this instance. Mods need a
/// loader; resource packs and shaders go into any instance.
async fn target_of(state: &AppState, instance_id: &str, kind: ProjectKind) -> Result<Target> {
    if kind == ProjectKind::Modpack {
        return Err(LauncherError::new(
            ErrorKind::Instance,
            "Модпак ставится как новая сборка, а не внутрь существующей",
        ));
    }
    let meta = instances::read_meta(&state.paths, instance_id).await?;
    if kind == ProjectKind::Mod && meta.loader == ModLoader::Vanilla {
        return Err(LauncherError::new(
            ErrorKind::Instance,
            "В ванильную сборку моды не ставятся — создайте сборку с лоадером",
        ));
    }
    Ok(Target::for_content(&meta.mc_version, meta.loader, kind))
}

#[tauri::command]
pub async fn available_providers(state: State<'_, AppState>) -> Result<Vec<ProviderId>> {
    Ok(mods::providers(&state.settings(), &state.client())
        .iter()
        .map(|provider| provider.id())
        .collect())
}

#[tauri::command]
pub async fn search_projects(state: State<'_, AppState>, query: SearchQuery) -> Result<SearchResult> {
    let provider = mods::provider(&state.settings(), &state.client(), query.provider)?;
    provider.search(&query).await
}

#[tauri::command]
pub async fn list_categories(
    state: State<'_, AppState>,
    provider: ProviderId,
    kind: ProjectKind,
) -> Result<Vec<Category>> {
    let provider = mods::provider(&state.settings(), &state.client(), provider)?;
    provider.categories(kind).await
}

#[tauri::command]
pub async fn project_details(
    state: State<'_, AppState>,
    provider: ProviderId,
    project_id: String,
) -> Result<ProjectDetails> {
    let provider = mods::provider(&state.settings(), &state.client(), provider)?;
    provider.details(&project_id).await
}

/// Project ids already in the instance, per provider — the browser shows
/// them as installed.
#[tauri::command]
pub async fn installed_projects(
    state: State<'_, AppState>,
    instance_id: String,
    provider: ProviderId,
    kind: ProjectKind,
) -> Result<Vec<String>> {
    let index = ModIndex::load(&state.paths, &instance_id, kind).await?;
    Ok(index.projects(provider).into_iter().collect())
}

/// Versions of a project for the version picker, newest first. By default
/// only ones that fit the instance; `any_game_version` drops the Minecraft
/// version filter (the loader filter for mods stays — those would not load).
#[tauri::command]
pub async fn list_versions(
    state: State<'_, AppState>,
    instance_id: String,
    provider: ProviderId,
    project_id: String,
    kind: ProjectKind,
    any_game_version: bool,
) -> Result<Vec<ModVersion>> {
    let mut target = target_of(&state, &instance_id, kind).await?;
    if any_game_version {
        target.mc_version.clear();
    }
    let provider_impl = mods::provider(&state.settings(), &state.client(), provider)?;
    let versions = provider_impl.versions(&project_id, &target).await?;
    Ok(versions.into_iter().filter(|v| target.accepts(v)).collect())
}

/// `version_id` installs that exact version instead of the best fit.
#[tauri::command]
pub async fn resolve_install(
    state: State<'_, AppState>,
    instance_id: String,
    provider: ProviderId,
    project_id: String,
    kind: ProjectKind,
    version_id: Option<String>,
) -> Result<ResolvedInstallPlan> {
    let target = target_of(&state, &instance_id, kind).await?;
    let provider_impl = mods::provider(&state.settings(), &state.client(), provider)?;
    let installed = ModIndex::load(&state.paths, &instance_id, kind)
        .await?
        .projects(provider);
    resolve::plan(
        provider_impl.as_ref(),
        &target,
        kind,
        &project_id,
        version_id.as_deref(),
        &installed,
    )
    .await
}

/// Downloads a confirmed plan. Runs as a task so it shows in the bottom bar,
/// and returns once the files are in place.
#[tauri::command]
pub async fn install_plan(
    state: State<'_, AppState>,
    instance_id: String,
    kind: ProjectKind,
    plan: ResolvedInstallPlan,
) -> Result<()> {
    // Validates the kind (no modpacks here) before anything is downloaded.
    target_of(&state, &instance_id, kind).await?;
    let task = state.tasks.start(
        TaskKind::InstallMod,
        plan.primary.name.clone(),
        Some(instance_id.clone()),
    );
    let mut versions = vec![plan.primary];
    versions.extend(plan.dependencies);

    let outcome = install::install_versions(
        &state.client(),
        &state.paths,
        &instance_id,
        kind,
        &versions,
        state.settings().download_concurrency(),
        Arc::clone(&task) as Arc<dyn Progress>,
    )
    .await;

    match &outcome {
        Ok(()) => task.finish_ok(),
        Err(error) => task.finish_err(error),
    }
    outcome
}

#[tauri::command]
pub async fn check_updates(
    state: State<'_, AppState>,
    instance_id: String,
    kind: ProjectKind,
) -> Result<Vec<ModUpdate>> {
    let target = target_of(&state, &instance_id, kind).await?;
    let providers = mods::providers(&state.settings(), &state.client());
    install::check_updates(&providers, &state.paths, &instance_id, kind, &target).await
}

#[tauri::command]
pub async fn apply_updates(
    state: State<'_, AppState>,
    instance_id: String,
    kind: ProjectKind,
    updates: Vec<ModUpdate>,
) -> Result<()> {
    target_of(&state, &instance_id, kind).await?;
    if updates.is_empty() {
        return Ok(());
    }
    let title = match updates.as_slice() {
        [single] => single.latest.name.clone(),
        _ => format!("Обновление: {} файлов", updates.len()),
    };
    let task = state.tasks.start(TaskKind::InstallMod, title, Some(instance_id.clone()));

    let outcome = install::apply_updates(
        &state.client(),
        &state.paths,
        &instance_id,
        kind,
        &updates,
        state.settings().download_concurrency(),
        Arc::clone(&task) as Arc<dyn Progress>,
    )
    .await;

    match &outcome {
        Ok(()) => task.finish_ok(),
        Err(error) => task.finish_err(error),
    }
    outcome
}

/// Deletes a resource pack or shader pack, keeping its index in step.
#[tauri::command]
pub async fn remove_content(
    state: State<'_, AppState>,
    instance_id: String,
    kind: ProjectKind,
    file_name: String,
) -> Result<()> {
    if kind == ProjectKind::Mod {
        return crate::instances::contents::remove_mod(&state.paths, &instance_id, &file_name).await;
    }
    crate::instances::contents::remove_pack(&state.paths, &instance_id, kind, &file_name).await
}
